lib/PuppeteerSharp/WaitTask.cs

```diff
@@ -183,84 +87,127 @@ public void Dispose()
 
-            if (exception == null &&
-                await _isolatedWorld.EvaluateFunctionAsync<bool>("s => !s", success)
-                    .ContinueWith(
-                        task => task.IsFaulted || task.Result,
-                        TaskScheduler.Default)
-                    .ConfigureAwait(false))
+                var success = await _poller.EvaluateFunctionHandleAsync("poller => poller.result()").ConfigureAwait(false);
+                _result.TrySetResult(success);
+                await TerminateAsync().ConfigureAwait(false);
+            }
+            catch (Exception ex)
             {
-                if (success != null)
+                var exception = GetBadException(ex);
+                if (exception != null)
                 {
-                    await success.DisposeAsync().ConfigureAwait(false);
+                    await TerminateAsync(exception).ConfigureAwait(false);
                 }
-
-                return;
             }
+        }
 
-            if (exception?.Message.Contains("Execution context was destroyed") == true)
+        internal async Task TerminateAsync(Exception exception = null)
+        {
+            // The timeout timer might call this method on cleanup
+            if (_terminated)
             {
-                _ = Rerun();
                 return;
             }
 
-            if (exception?.Message.Contains("Cannot find context with specified id") == true)
-            {
-                return;
-            }
+            _terminated = true;
+            _isolatedWorld.TaskManager.Delete(this);
+            Cleanup(); // This matches the clearTimeout upstream
 
             if (exception != null)
             {
-                _taskCompletion.TrySetException(exception);
+                _result.TrySetException(exception);
             }
-            else
+
+            if (_poller != null)
             {
-                _taskCompletion.TrySetResult(success);
-            }
+                try
+                {
+                    await _poller.EvaluateFunctionAsync(@"async poller => {
+                        await poller.stop();
+                    }").ConfigureAwait(false);
 
-            Cleanup();
+                    await _poller.DisposeAsync().ConfigureAwait(false);
+                    _poller = null;
+                }
+                catch (Exception)
+                {
+                    // swallow error.
+                }
+            }
         }
 
-        internal void Terminate(Exception exception)
+        private Exception GetBadException(Exception exception)
         {
-            _terminated = true;
-            _taskCompletion.TrySetException(exception);
-            Cleanup();
+            if (exception.Message.Contains("Execution context is not available in detached frame"))
+            {
+                return new PuppeteerException("Waiting failed: Frame detached");
```

Comment:
Should `exception` be set as the inner exception of the constructed `PuppeteerException`?

Reply:
Yes!